cli/command/image/build.go

```diff
@@ -373,6 +478,86 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
 	return nil
 }
 
+type sizeProgress struct {
+	out     progress.Output
+	action  string
+	limiter *rate.Limiter
+}
+
+func (sp *sizeProgress) update(size int, last bool) {
+	if sp.limiter == nil {
+		sp.limiter = rate.NewLimiter(rate.Every(100*time.Millisecond), 1)
+	}
+	if last || sp.limiter.Allow() {
+		sp.out.WriteProgress(progress.Progress{Action: sp.action, Current: int64(size), LastUpdate: last})
+	}
+}
+
+type bufferedWriter struct {
+	done chan error
+	io.Writer
+	buf     *bytes.Buffer
+	flushed chan struct{}
+	mu      sync.Mutex
+}
+
+func newBufferedWriter(done chan error, w io.Writer) *bufferedWriter {
+	bw := &bufferedWriter{done: done, Writer: w, buf: new(bytes.Buffer), flushed: make(chan struct{})}
+	go func() {
+		<-done
+		bw.flushBuffer()
+	}()
+	return bw
+}
+
+func (bw *bufferedWriter) Write(dt []byte) (int, error) {
+	select {
+	case <-bw.done:
+		bw.flushBuffer()
+		return bw.Writer.Write(dt)
+	default:
+		return bw.buf.Write(dt)
+	}
+}
+
+func (bw *bufferedWriter) flushBuffer() {
+	bw.mu.Lock()
+	select {
+	case <-bw.flushed:
+	default:
+		bw.Writer.Write(bw.buf.Bytes())
+		close(bw.flushed)
+	}
+	bw.mu.Unlock()
+}
+
+func getBuildSharedKey(dir string) (string, error) {
+	dt := []byte(cliconfig.Dir()) // if no access to config dir then only use path
+	if err := os.MkdirAll(cliconfig.Dir(), 0700); err == nil {
+		// build session is hash of build dir with node based randomness
+		sessionFile := filepath.Join(cliconfig.Dir(), ".buildsharedkey")
+		if _, err := os.Lstat(sessionFile); err != nil {
+			if os.IsNotExist(err) {
+				b := make([]byte, 32)
+				if _, err := rand.Read(b); err != nil {
+					return "", err
+				}
+				if err := ioutil.WriteFile(sessionFile, []byte(hex.EncodeToString(b)), 0600); err != nil {
+					return "", err
+				}
+			}
+		}
+
+		dt, err = ioutil.ReadFile(sessionFile)
```

Comment:
There will never be multiple instances of sharedkey and create is only called if read fails. Maybe helps to split this to `getSessionSharedKey(dir)` and `getNodeIdentifier()` as some hashing is always needed. Getting node identifier can be skipped if it has already run once.